### PR TITLE
Added GemJoin8 and GUSD example

### PR DESCRIPTION
**Pull request by hexonaut**
Added GemJoin8 to support GUSD. To improve on re-usability for proxy contracts, I've allowed for arbitrary function names which return the current implementation.

### CODE REVIEW

File: src/join.sol [hunk 1]
```diff
@@ -502,3 +502,76 @@ contract GemJoin7 is LibNote {
...
+    mapping (address => uint) public wards;
+    function rely(address usr) external note auth { wards[usr] = 1; }
+    function deny(address usr) external note auth { wards[usr] = 0; }
+    modifier auth { require(wards[msg.sender] == 1); _; }
+
+    VatLike  public vat;
+    bytes32  public ilk;
+    GemLike8 public gem;
+    bytes    public funcSig;
+    uint     public dec;
```
Comment (brianmcmichael) [inline src/join.sol]: Recommend explicit `uint256` declaration wherever applicable.

File: src/join.sol [hunk 2]
```diff
@@ -502,3 +502,76 @@ contract GemJoin7 is LibNote {
...
+
+    constructor(address vat_, bytes32 ilk_, address gem_, string memory func_) public {
+        gem = GemLike8(gem_);
+        dec = gem.decimals();
+        require(dec < 18, "GemJoin8/decimals-18-or-higher");
+        wards[msg.sender] = 1;
+        live = 1;
+        vat = VatLike(vat_);
+        ilk = ilk_;
+        funcSig = abi.encodeWithSignature(func_);
```
Comment (brianmcmichael) [inline src/join.sol]: Interesting solution for alternate implementation checks. This constructor could benefit from some documentation of the parameters to ensure that it's the implementation call that's getting passed here, and not, say, a `transfer()` function. Using a name like `implFunc_` might also help clarify.

File: src/tokens.sol [hunk 3]
```diff
@@ -1044,3 +1044,72 @@ contract LINK {
...
+    }
+    function sub(uint x, uint y) internal pure returns (uint z) {
+        require((z = x - y) <= x, "ds-math-sub-underflow");
+    }
+    event Approval(address indexed src, address indexed guy, uint wad);
+    event Transfer(address indexed src, address indexed dst, uint wad);
+
+    string  public  name = "GUSD";
+    string  public  symbol = "GUSD";
+    uint256  public  decimals = 2;
```
Comment (brianmcmichael) [inline src/tokens.sol]: The decimals on GUSD returns a `uint8`

**Comment by gbalabasquer** [comment]
@hexonaut sorry for the inconvenience but would you mind migrating this PR changes to https://github.com/makerdao/dss-gem-joins?
We have just extracted all the tokens and adapters to that repo.

**Comment by hexonaut** [comment]
Sure thing.

**Comment by gbalabasquer** [comment]
Thank you! Closing this PR.

**Pull request closed without merging**